src/Devices/EthernetServer_TC.cpp

```diff
@@ -419,29 +429,60 @@ void EthernetServer_TC::sendHeadersWithSize(uint32_t size) {
   // blank line indicates end of headers
   client.write('\r');
   client.write('\n');
-  state = FINISHED;  // TODO: Why?! This is awkward when we want to send a body next.
 }
 
-// 303 response
-void EthernetServer_TC::sendRedirectHeaders() {
-  static const char response[] PROGMEM =
+void EthernetServer_TC::sendResponse(int code) {
+  static const char response_303[] PROGMEM =
       "HTTP/1.1 303 See Other\r\n"
       "Location: /api/1/display\r\n"
       "Access-Control-Allow-Origin: *\r\n"
       "\r\n";
-  char buffer[sizeof(response)];
-  strncpy_P(buffer, (PGM_P)response, sizeof(buffer));
-  client.write(buffer);
-  state = FINISHED;
-}
-
-// 400 response
-void EthernetServer_TC::sendBadRequestHeaders() {
-  char buffer[30];
-  static const char response[] PROGMEM = "HTTP/1.1 400 Bad Request\r\n\r\n";
-  strncpy_P(buffer, (PGM_P)response, sizeof(buffer));
+  static const char response_400[] PROGMEM =
+      "HTTP/1.1 400 Bad Request\r\n"
+      "\r\n";
+  static const char response_404[] PROGMEM =
+      "HTTP/1.1 404 Not Found\r\n"
+      "\r\n";
+  static const char response_405[] PROGMEM =
+      "HTTP/1.1 405 Method Not Allowed\r\n"
+      "Allow: GET, POST\r\n"
+      "\r\n";
+  static const char response_408[] PROGMEM =
+      "HTTP/1.1 408 Request Timeout\r\n"
+      "Connection: close\r\n"
+      "\r\n";
+  static const char response_500[] PROGMEM =
+      "HTTP/1.1 500 Internal Server Error\r\n"
+      "\r\n";
+  static const char response_501[] PROGMEM =
+      "HTTP/1.1 501 Not Implemented\r\n"
+      "\r\n";
+  static const char *const response_table[] PROGMEM = {response_303, response_400, response_404, response_405,
+                                                       response_408, response_501, response_500};
+  char buffer[sizeof(response_303)];
+  switch (code) {
+    case HTTP_REDIRECT:
+      strncpy_P(buffer, (char *)pgm_read_word(&(response_table[0])), sizeof(buffer));
```

Comment:
Try `(PGM_P)response_303` instead of `&(response_table[0])` (and equivalent adjustments below, removing the `response_table`).

Reply:
Thanks, I've made these changes and it helped with the tests.